marketo_ma.module

```diff
@@ -538,15 +527,64 @@ function _marketo_ma_save_lead_fields_database($fields) {
  * @return array
  *   Associative array of lead fields keyed by their REST API name
  */
-function _marketo_ma_get_lead_fields_database() {
+function _marketo_ma_get_lead_fields_database($key = 'id') {
   $result = db_select('marketo_ma_lead_fields', 'f')
     ->fields('f')
     ->orderBy('displayName')
     ->execute()
-    ->fetchAllAssoc('restName', PDO::FETCH_ASSOC);
+    ->fetchAllAssoc($key, PDO::FETCH_ASSOC);
   return $result;
 }
 
+/*
+ * Ensures fields are mapped to the proper key based on tracking method.
+ *
+ * @param array $lead
+ *   Lead
+ * @param string $tracking_method
+ *   Marketo tracking method
+ */
+function _marketo_ma_resolve_field_names(&$lead, $tracking_method) {
+  if ($tracking_method == 'rest' || $tracking_method == 'rest_async') {
+    $key = 'restName';
+  }
+  else {
+    $key = 'soapName';
+  }
+  $fields = _marketo_ma_get_lead_fields_database();
+  $lead_data = array();
+  // Sort data so text keys are first. Text keys were likely set manually rather
+  // than by mapped fields and we'll let them be overwritten by mappings.
+  ksort($lead['data']);
+  foreach ($lead['data'] as $label => $value) {
+    if (is_numeric($label) && array_key_exists($label, $fields)) {
+      // Change the numeric ID to the field key
+      $lead_data[$fields[$label][$key]] = $value;
+    }
+    else if (is_null($label) || trim($label) == '') {
+      // Shouldn't happen but ignore anything with a blank key
+      watchdog('marketo', 'A blank lead key was found, data will not be submitted. !label = !value ', array('!label' => $label, '!value' => $value), WATCHDOG_WARNING);
+    }
+    else {
+      // Check the DB to see if we can match on other fields
+      $result = db_select('marketo_ma_lead_fields', 'f')
+        ->fields('f')
+        ->condition(db_or()->condition('restName', $label)->condition('soapName', $label))
```

Comment:
That makes sense, thanks!

>hook_marketo_ma_lead_FIELDNAME_alter needs to be reviewed also as the change to ID will impact any existing implementations.

Do you mean we are using this in the module directly, or in contrib / custom modules using this implementation at present?

On that note, will we provide an upgrade patch / update hook for what was there previously? It's not a requirement to do either as 2.x is still a dev release.

Will you not change the schema to map on `marketo_id` in the *marketo_ma_webform_component* table? If not, what if the tracking method is changed for some site down the line, will this not break their webforms integration with Marketo?